pkg/core/blockchain.go

```diff
@@ -1218,24 +1152,29 @@ func (bc *Blockchain) GetStandByValidators() (keys.PublicKeys, error) {
 // GetValidators returns validators.
 // Golang implementation of GetValidators method in C# (https://github.com/neo-project/neo/blob/c64748ecbac3baeb8045b16af0d518398a6ced24/neo/Persistence/Snapshot.cs#L182)
 func (bc *Blockchain) GetValidators(txes ...*transaction.Transaction) ([]*keys.PublicKey, error) {
-	chainState := NewBlockChainState(bc.store)
 	if len(txes) > 0 {
 		for _, tx := range txes {
 			// iterate through outputs
 			for index, output := range tx.Outputs {
 				accountState := bc.GetAccountState(output.ScriptHash)
-				accountState.Balances[output.AssetID] = append(accountState.Balances[output.AssetID], UnspentBalance{
+				accountState.Balances[output.AssetID] = append(accountState.Balances[output.AssetID], entities.UnspentBalance{
 					Tx:    tx.Hash(),
 					Index: uint16(index),
 					Value: output.Amount,
 				})
+				if err := bc.dao.PutAccountState(accountState); err != nil {
+					return nil, err
+				}
 				if output.AssetID.Equals(governingTokenTX().Hash()) && len(accountState.Votes) > 0 {
 					for _, vote := range accountState.Votes {
-						validatorState, err := chainState.validators.getAndUpdate(chainState.store, vote)
+						validatorState, err := bc.dao.GetValidatorStateOrNew(vote)
```

Comment:
You should have temporary `dao` here also, this is not the place where you change blockchain's state.

Reply:
used `cache` as in `storeBlock()`